storage/http_client.go

```diff
@@ -1468,18 +1477,64 @@ func parseReadResponse(res *http.Response, params *newRangeReaderParams, reopen
 		CRC32C:          crc,
 		Decompressed:    res.Uncompressed || uncompressedByServer(res),
 	}
-	return &Reader{
+
+	httpReader := &httpReader{
+		reopen:   reopen,
+		body:     body,
+		wantCRC:  crc,
+		checkCRC: checkCRC,
+		closed:   make(chan bool),
+	}
+	r := &Reader{
 		Attrs:    attrs,
 		size:     size,
 		remain:   remain,
 		checkCRC: checkCRC,
-		reader: &httpReader{
-			reopen:   reopen,
-			body:     body,
-			wantCRC:  crc,
-			checkCRC: checkCRC,
-		},
-	}, nil
+		reader:   httpReader,
+		ctx:      ctx,
+	}
+
+	// If retryOption minReadThroughput was provided, monitor the throughput of the read once
+	// per second and, if the throughput is below the minimum, close the read body to prompt
+	// reopening the stream.
+	var monitorThroughput func()
+
+	if s.retry != nil && s.retry.minReadThroughput > 0 {
+		monitorThroughput = func() {
+			var readCalls int
+			var seen int64
+			var done bool
+			for !done {
+				select {
+				case <-time.After(time.Second):
+					// Only close here if a Read call is in progress or if there have been calls
+					// to Read in the past second. This avoids needlessly closing the stream in
+					// the case where the application is not making calls to Read and so
+					// throughput is low for that reason.
+					if (httpReader.readCalls-readCalls > 1 || httpReader.readCallOpen) && httpReader.seen-seen < s.retry.minReadThroughput {
```

Comment:
`httpReader.readCalls` is not thread safe, do we need to add a mutex around it?

Reply:
Good point -- I did this. There might be a cleaner way to accomplish this with a channel however.